streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -271,30 +279,43 @@ private long processNext(final List<ConsumerRecord<byte[], byte[]>> records,
                              final StateRestorer restorer,
                              final Long endOffset) {
         final List<KeyValue<byte[], byte[]>> restoreRecords = new ArrayList<>();
-        long offset = -1;
-
+        long nextPosition = -1;
+        int numberRecords = records.size();
+        int numberRestored = 0;
         for (final ConsumerRecord<byte[], byte[]> record : records) {
-            offset = record.offset();
+            final long offset = record.offset();
             if (restorer.hasCompleted(offset, endOffset)) {
+                nextPosition = record.offset();
                 break;
             }
+            numberRestored++;
             if (record.key() != null) {
                 restoreRecords.add(KeyValue.pair(record.key(), record.value()));
             }
         }
 
-        if (offset == -1) {
-            offset = restoreConsumer.position(restorer.partition());
+        if (nextPosition == -1) {
+            nextPosition = restoreConsumer.position(restorer.partition());
         }
 
         if (!restoreRecords.isEmpty()) {
             restorer.restore(restoreRecords);
-            restorer.restoreBatchCompleted(offset + 1, records.size());
+            restorer.restoreBatchCompleted(nextPosition, records.size());
+
+        }
+
+        // if it's changelog topic and we did not restore all records in list, need to return last offset contained in list
+        // to force a TaskMigratedException in restorePartition method, as we encountered race condition and the
+        // changelog topic has been updated
+        if (restorer.isChangeLogTopic() && numberRestored != numberRecords) {
+            nextPosition = records.get(numberRecords - 1).offset();
```

Comment:
I think we can simplify this and remove this check completely... I simple `return Max.min(nextPosition, restorer.getOffsetLimit)` below might be sufficient.

Reply:
I tried that, but the test `shouldThrowTaskMigratedExceptionIfEndOffsetGetsExceededDuringRestoreForChangelogTopic` fails as we return `5` for `nextPosition` but the max offset fetched is `10`  

cf previous comment https://github.com/apache/kafka/pull/4300#discussion_r157029161

Reply:
Did you remove `nextPosition = record.offset();` in L288 -- guess we don't need this anymore if we return `Math.min(...)`

Reply:
Ack, Now I recall the full discussion.